db/hints/manager.cc

```diff
@@ -286,36 +287,30 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
     }
```

Comment:
> This commit changes the range we iterate through.
Instead of iterating over endpoint managers, we start
iterating over shard_rps. The rationale for that is
that usually, the size of shard_rps should be smaller
than the number of endpoint managers currently
managed by the hint shard manager. Also, if no endpoint
manager is found, we immediately return.

If this is pure waiting, then there are no negative side effects. However, if wait_until_...() also kicks something, then we may delay kicking it.

On the other hand, if it's pure, why does changing the loop nesting matter?

Reply:
It is pure waiting indeed. The rationale for changing the range we loop over was to handle situations when we might have hundreds of endpoint managers, but there are just a few replay positions in the `shard_rps` that is passed to the function. Instead of creating hundreds of fibers, we would have only a few. Of course, the cost shouldn't be too big in either case, so it was more of a microoptimization rather than something having visible effect on performance.

However, it turns out that I must've been tired when refactoring the code because this specific commit changed the semantics of the function and I didn't notice it. Segments that were moved during hint rebalancing are treated differently; it more or less corresponds to the replay positions of those "foreign" segments being equal to minus infinity. Because of that, even if there is no replay position for a specific endpoint, we might still need to wait for the corresponding endpoint manager. As a result, iterating over all endpoint managers is inevitable.

Thank you for this comment. I'll go back to the previous semantics.